File: java/otterop/lang/src/main/java/otterop/lang/Array.java

package otterop.lang;

public class Array<T> {
    private T[] _wrapped;

    private Array(T[] array) {
        this._wrapped = array;
    }

    public T get(int i) {
        return _wrapped[i];
    }

    public void set(int i, T value) {
        _wrapped[i] = value;
    }

    public int size() {
        return _wrapped.length;
    }

    public static <T> Array<T> wrap(T[] list) {
        return new Array<T>(list);
    }

    public static Array<String> wrapString(java.lang.String[] arg) {
        String[] ret = new String[arg.length];
        for (int i = 0; i < arg.length; i++) {
            ret[i] = String.wrap(arg[i]);
        }
        return new Array<>(ret);
    }
}

File: dotnet/Lang/Array.cs
namespace Otterop.Lang;
public class Array
{
    internal Array() { }

    public static Array<String> WrapString(string[] args)
    {
        String[] wrappedStrings = new String[args.Count()];
        int i = 0;
        foreach(var arg in args)
        {
            wrappedStrings[i] = String.Wrap(arg);
            i++;
        }
        return new Array<String>(wrappedStrings, 0, wrappedStrings.Count());
    }
}

public class Array<T> : Array
{
    private T[] wrapped;

    private int start;

    private int end;

    internal Array(T[] wrapped, int start, int end)
    {
        this.wrapped = wrapped;
        this.start = start;
        this.end = end;
    }

    public T Get(int i)
    {
        return wrapped[start + i];
    }

    public void Set(int i, T value)
    {
        wrapped[start + i] = value;
    }

    public Array<T> Slice(int start, int end) {
        var newStart = this.start + start;
        var newEnd = this.start + end;
        if (newStart < this.start || newStart > this.end || newEnd < newStart ||
            newEnd > this.end) {
                throw new IndexOutOfRangeException();
        }
        return new Array<T>(wrapped, newStart, newEnd);
    }

    public int Size()
    {
        return end - start;
    }
}